library(shiny)
library(dplyr)
library(plotly)
library(cBioPortalData)
library(DT)
library(stringr)
options(dplyr.summarise.inform = FALSE)

sample <- "CPDC181710"
sample_path <- paste0("../../CPDC181710/")

karyotype_filename <- paste0(sample, ".cnv.pdf")
karyotype_file <- paste0(sample_path, karyotype_filename)
file.copy(karyotype_file, "www")

cnr_file <- read.table(paste0(sample_path, sample, ".final.cnr"), sep = '\t', header = TRUE)
cnr_target <- filter(cnr_file, !gene %in% c("Antitarget", ".")) %>% select(chromosome, gene)
genes <- c("", sort(unique(cnr_target$gene)))

adjusted_gene_file <- read.csv(paste0(sample_path, sample, "_genes.csv"))
adj_genes <- c("", sort(unique(adjusted_gene_file$gene.symbol)))

cbio <- cBioPortal()
cbio_studies <- read.csv("cbio_study_names.csv")

chrs <- c("all", paste0("chr", "1":"22"), "chrX", "chrY")

fluidPage(
  
  navbarPage("CNViz",
             tabPanel("Patient Data (adjusted)", fluid=TRUE,
                      sidebarLayout(
                        sidebarPanel(
                          width = 3,
                          selectInput(inputId = "adj_chr",
                                      label = "chromosome",
                                      choices = chrs,
                                      selected = "all"),
                          selectizeInput(inputId = "adj_gene",
                                         label = "gene",
                                         choices = adj_genes,
                                         selected = ""),
                          br(), 
                          img(src = "../adjusted_legend.jpg", width = "100%")
                        ),
                        mainPanel(
                          h3(sample),
                          tableOutput("meta"),
                          textOutput("comment"),
                          br(),
                          plotlyOutput("adj_chr_plot", width = "100%"),
                          textOutput("adj_copies"),
                          dataTableOutput("mutations"),
                          br(),
                          conditionalPanel("input.adj_chr != 'all'", plotlyOutput("adj_selected_plot"))
                        )
                      )),
             tabPanel("Patient Data (unadjusted)", fluid=TRUE, 
                      sidebarLayout(
                        sidebarPanel(
                          width = 3,
                          selectInput(inputId = "chr",
                                      label = "chromosome",
                                      choices = chrs,
                                      selected = "all"),
                          selectizeInput(inputId = "gene",
                                         label = "gene",
                                         choices = genes,
                                         selected = ""),
                          a("Karyotype",target="_blank",href=karyotype_filename),
                          br(), br(), 
                          img(src = "../unadjusted_legend.jpg", width = "100%"),
                          br(), br()
                        ),
                        mainPanel(
                          width = 9,
                          h3(sample),
                          plotlyOutput("chr_plot", width = "100%"),
                          br(),
                          conditionalPanel("input.chr != 'all'", plotlyOutput("selected_plot"))
                        )
                      )),
             tabPanel("TCGA Pan-Cancer Atlas 2018 Data", fluid=TRUE,
                      selectizeInput(inputId = "cancer", label = "cancer",
                                     choices = c("", cbio_studies$Cancer),
                                     selected = ""),
                      dataTableOutput("cbioOutput")),
             tabPanel(icon("info-circle", class = NULL, lib = "font-awesome"), fluid=TRUE,
                      h5("Patient Data (Unadjusted)"),
                      p("Unadjusted data is obtained from CNVkit. CNVkit takes hybrid capture DNA sequencing data and infers copy number. Copy number estimates generated by CNVKit are not necessarily integers. This tab also includes probe-level data for each gene."),
                      p(em("Talevich, E., Shain, A.H., Botton, T., & Bastian, B.C. (2014). CNVkit: Genome-wide copy number detection and visualization from targeted sequencing. PLOS Computational Biology 12(4):e1004873.")),
                      div(style="display: inline-block;", a("CNVkit paper,", target ="_blank", href = "https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004873")),
                      div(style="display: inline-block;", a("CNVkit documentation", target ="_blank", href = "https://cnvkit.readthedocs.io/en/stable/")),
                      br(), br(),
                      h5("Patient Data (Adjusted)"),
                      p("Adjusted data is obtained from PureCN, using data generated by CNVkit. PureCN estimates tumor purity, tumor ploidy, loss of heterozygosity (LOH) and integer copy number. Somatic mutations are also inferred. Of note, germline mutations are not displayed. Tumor purity is PureCN's estimate for the percentage of the sample containing tumor cells. Ploidy is estimated by PureCN, then rounded to the nearest whole number. Only segments where copy number differs from tumor ploidy or those with LOH are highlighted. If a sample is marked as Flagged, this means PureCN has deemed the results as unreliable. The unit displayed on the x-axis for the adjusted data is chromosome band locations. See unadjusted view for megabase location and probe-level data."),
                      p(em("Riester, M., Singh, A.P., Brannon, A.R. et al. (2016). PureCN: copy number calling and SNV classification using targeted short read sequencing. Source Code Biol Med 11(1), 13.")),
                      div(style="display: inline-block;", a("PureCN paper,", target ="_blank", href = "https://scfbm.biomedcentral.com/articles/10.1186/s13029-016-0060-z")),
                      div(style="display: inline-block;", a("PureCN documentation", target ="_blank", href = "https://rdrr.io/bioc/PureCN/f/inst/doc/PureCN.pdf")),
                      br(), br(),
                      h5("TCGA Pan-Cancer Atlas Data"),
                      p("TCGA Pan-Cancer Atlas Data was obtained from cBioPortal's R package cbioportaldata. This same information can be found on cBioPortal's website. Similar information from many additional studies can also be found on their website. The copy number data displayed was generated by the GISTIC algorithm."),
                      p(" - Deep Deletion indicates a deep loss, possibly a homozygous deletion"),
                      p(" - Shallow Deletion indicates a shallow loss, possibley a heterozygous deletion"),
                      p(" - Gain indicates a low-level gain (a few additional copies, often broad)"),
                      p(" - Amplification indicate a high-level amplification (more copies, often focal)"),
                      div(style="display: inline-block;", a("cBioPortal website,", target ="_blank", href = "https://www.cbioportal.org/")),
                      div(style="display: inline-block;", a("cBioPortal FAQ,", target ="_blank", href = "https://docs.cbioportal.org/1.-general/faq")),
                      div(style="display: inline-block;", a("cbioportaldata R Package,", target ="_blank", href = "https://bioconductor.org/packages/release/bioc/html/cBioPortalData.html")),
                      div(style="display: inline-block;", a("GISTIC paper", target ="_blank", href = "https://pubmed.ncbi.nlm.nih.gov/18077431/"))
                      ))
)